### DISCUSSION TITLE
reference to the FEProblem

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#22031] [Q&A General] [answered]
My volume score inherits **_ArrayKernel.C_** and area score inherits ArrayDGKernel.C. I want to use a reference to the FEProblem class. **_So I wrote the statement "_fe_problem(*parameters.get<FEProblem *>("_fe_problem"))"_** in the constructor of the subclass and successfully used it to fetch something from the FEProblem class, but this statement causes my configuration file to not work properly, i.e. it compiles successfully but does not run the configuration file. I would like to know if there is any way to compile and run it successfully.

**Comment by GiudGiud** [comment, marked as answer]
I don’t think FEProblem is a parameter the user passes. 
there’s examples in the framework for retrieving the FEProblem, you should grep for it

**Reply by xiaojbing** [reply]
In fact, I want to get the parameters in a problem file I wrote myself in the subclass of ArrayDGKernel. And the problem file I wrote myself inherits FEProblem.cpp

**Reply by GiudGiud** [reply]
You wrote your own problem class? and problem is a parameter in the input file, as in you have two problems active?

So the way this is done, and you can look at base/ResidualObject.C for example, is to retrieve the problem through a pointer to the problem. The parameter is a pointer. This pointer is set internally, not in the input file. 
```
_fe_problem(*parameters.get<FEProblem *>("_fe_problem")),
```

**Comment by xiaojbing** [comment]
I found that _fe_problem can be called in the file that inherits ArrayKernel.C, but _fe_problem cannot be called in the file that inherits ArrayDGKernel.C. I searched for _c_fe_problem,_mci_feproblem,_mi_feproblem,_ sc_fe_problem,_ti_feproblem, what is the difference between these and _fe_problem, and if they have the same meaning, can I use them?

**Reply by GiudGiud** [reply]
_fe_problem can be called in the classes that have this attribute. 

See the headers or the doxygen for each attribute. 
For example in the header of the Coupleable interface, you can see that _c_fe_problem is a reference to the FEProblemBase (a base class of FEProblem)
